ndk/src/main/jni/bugsnag_ndk.c

```diff
@@ -0,0 +1,422 @@
+void bsg_release_env_write_lock(void) {
+  pthread_mutex_unlock(&bsg_global_env_write_mutex);
+}
+
+bsg_unwinder bsg_configured_unwind_style() {
+  if (bsg_global_env != NULL)
+    return bsg_global_env->unwind_style;
+
+  return BSG_CUSTOM_UNWIND;
+}
+
+JNIEXPORT void JNICALL Java_com_bugsnag_android_ndk_NativeBridge_install(
+    JNIEnv *env, jobject _this, jstring _report_path, jboolean auto_notify,
+    jint _api_level) {
+  bsg_environment *bugsnag_env = calloc(1, sizeof(bsg_environment));
+  bugsnag_env->unwind_style = bsg_get_unwind_type((int)_api_level);
+  bugsnag_env->report_header.big_endian =
+      htonl(47) == 47; // potentially too clever, see man 3 htonl
+  bugsnag_env->report_header.version = BUGSNAG_REPORT_VERSION;
+  const char *report_path = (*env)->GetStringUTFChars(env, _report_path, 0);
+  sprintf(bugsnag_env->next_report_path, "%s", report_path);
+
+  if ((bool)auto_notify) {
+    bsg_handler_install_signal(bugsnag_env);
+  }
+
+  // populate metadata from Java layer
+  bsg_populate_report(env, &bugsnag_env->next_report);
+
+  // If set, save os build info to report info header
+  if (strlen(bugsnag_env->next_report.device.os_build) > 0) {
+    bsg_strncpy_safe(bugsnag_env->report_header.os_build,
+                     bugsnag_env->next_report.device.os_build,
+                     sizeof(bugsnag_env->report_header.os_build));
+  }
+
+  (*env)->ReleaseStringUTFChars(env, _report_path, report_path);
+  bsg_global_env = bugsnag_env;
+  BUGSNAG_LOG("Initialization complete!");
+}
+
+JNIEXPORT void JNICALL
+Java_com_bugsnag_android_ndk_NativeBridge_deliverReportAtPath(
+    JNIEnv *env, jobject _this, jstring _report_path) {
+  static pthread_mutex_t bsg_native_delivery_mutex = PTHREAD_MUTEX_INITIALIZER;
+  pthread_mutex_lock(&bsg_native_delivery_mutex);
+  const char *report_path = (*env)->GetStringUTFChars(env, _report_path, 0);
+  bugsnag_report *report =
+      bsg_deserialize_report_from_file((char *)report_path);
+
+  if (report != NULL) {
+    char *payload = bsg_serialize_report_to_json_string(report);
+    if (payload != NULL) {
+      jclass interface_class =
+          (*env)->FindClass(env, "com/bugsnag/android/NativeInterface");
+      jmethodID jdeliver_method =
+          (*env)->GetStaticMethodID(env, interface_class, "deliverReport",
+                                    "(Ljava/lang/String;Ljava/lang/String;)V");
+      jstring jpayload = (*env)->NewStringUTF(env, payload);
+      jstring jstage = (*env)->NewStringUTF(env, report->app.release_stage);
+      (*env)->CallStaticVoidMethod(env, interface_class, jdeliver_method,
+                                   jstage, jpayload);
+      (*env)->DeleteLocalRef(env, jpayload);
+      (*env)->DeleteLocalRef(env, jstage);
+      free(payload);
+    } else {
+      BUGSNAG_LOG("Failed to serialize report as JSON: %s", report_path);
+    }
+    free(report);
+  } else {
+    BUGSNAG_LOG("Failed to read report at file: %s", report_path);
+  }
+  remove(report_path);
```

Comment:
Correct. In this case, one of three things occurred:

1. The report was deserialized correctly. It will then be passed synchronously to the Java layer to be delivered. In the worst case, say the app crashes while this process is incomplete, the file won't be deleted.
2. The file did not exist
3. It could not be read, and thus is deleted since nothing can be done with it. This is the case which logs a warning.